### DISCUSSION TITLE
[Question] Iterating on Array in Configured API Request

### DISCUSSION THREAD

**Discussion by julroger2013** [Rooyca/obsidian-api-request#16] [General]
I had a question about the configured API requests. 

I have a codeblock formatted as so:

```req
url: https://api.todoist.com/rest/v2/tasks?filter=`contentOfFilter`
method: get
headers: { "Authorization": "Bearer myToken"}
show: {..} -> content
```
Todoist returns an array of JSON objects, with each object representing a task, with various components such as id, content, priority, and so on.

This codeblock works perfectly in extracting the "content" component of each JSON object in the array - functionally, it's grabbing the name of each task that Todoist sends back.

I'm finding that I cannot figure out how to obtain the same results with the preconfigured API requests that you can set up in the plugin settings.

I would imagine that limiting to something like "content" - one component of each object in an array of objects - is done with the "What to display" box in the plugin settings, but I can't seem to figure out exactly how to set up the preconfigured API requests to obtain the same result as the codeblock above.

Any ideas on how I need to structure this configured request to obtain the desired result?

Thank you!

**Comment by Rooyca** [comment]
Hey @julroger2013!
In the last update I was planning to remove that part of the plugin, it seemed to me that it didn't add much and was a bit convoluted. Glad to hear someone is using it.

The short answer to your question is: **You can't**.
The codeblock part and the configuration part are two separate parts.

If it's not too forward I would like to know why not to use codeblocks?

**Comment by julroger2013** [comment]
Good question. Maybe you can help me figure out how to achieve the desired result I have in mind, whether that uses codeblocks or configured API requests.

I manage my tasks with Todoist, which has a pretty well-established REST API. 

What I'm wanting to do is pull a list of tasks from Todoist in markdown format using the API. Each task is a JSON object with many components, such as priority, due date, UID, task name. 

I'd like to pull tasks using a filter function built into the API, and then extract only the task names from the returned JSON objects.

I would want to format these as a markdown checklist, e.g. ...

- [ ] taskName 1
- [ ] taskName 2
- [ ] taskName 3
- [ ] taskName 4

And so on. That list then goes through a script (which I already have set up through Templater) which pairs the tasks against each other head-to-head (i.e. 1 v. 2, 1 v. 3, 2 v. 3, etc.) to prioritize. The task that ends up with the most head-to-head victories ends up first priority, second most as second priority, and so on.

Basically, I'm just trying to use your plugin to manage the first part of this workflow - extracting the tasks from Todoist's API and putting them into a markdown checklist that can be processed by the script that I've already established. I figured I could use the "Paste Result" command from your plugin to produce that list in one step.

Does that make sense? I'm open-minded about how best to approach this problem. 

I'm trying to get as close as I can to a single button-press to extract tasks, format them in a markdown checklist, and run them through the prioritization process.

Thank you!

**Comment by Rooyca** [comment]
I think I'll have to change or add a couple of things on the plugin for this to work the way you need to.

When I find a solution I'll write you.

**Comment by Rooyca** [comment]
I found a 'easy' way of doing this:

```req
url: https://api.todoist.com/rest/v2/tasks
headers: {"Authorization": "Bearer TOKEN"}
show: {..} -> content
format: <input type="checkbox"> {} </input> <br>
```

Do you think that would work with your workflow?

**Comment by julroger2013** [comment]
The format is pretty much right. The one downside is that it's not possible to drag-select the checkboxes, too, for copy-paste purposes. That's why I hoped to instead format things in a configured request and paste them with a command.

As I'm sitting here thinking about it, though ... Templater can run JavaScript - is there any reason I can't just do an API request as part of a Templater script and then use other JavaScript functions as part of that script to format the returned values?

Thanks for thinking this over!

**Comment by julroger2013** [comment]
One possibility - would it be hard to add a "copy contents" button to the top right corner of your code blocks? They have those in ordinary code blocks & also optionally in callouts, so you can copy the contents of the block. That might work well, and maybe would be easier than the configured portion of your plugin?

Thanks again!

**Comment by Rooyca** [comment]
> Templater can run JavaScript 

I believe you could do that. Don't  know if it works tho.

>  add a "copy contents" button

I did that. Would be live in the next update. 
For this to work you should remove the `<input>` an replace it with a `- [ ]`.

```req
url: https://api.todoist.com/rest/v2/tasks
headers: {"Authorization": "Bearer TOKEN"}
show: {..} -> content
format: - [ ] {} <br>
```

**Comment by julroger2013** [comment]
Great, thank you!

A marginal idea, since the main benefit will already come from the copy contents function, but do you suppose your code blocks could be made to render markdown?

The Better Search Views plugin probably has useful code when it comes to markdown rendering. It makes markdown render in, eg, backlinks views. 

Thanks again for adding the copy contents button! That will get me where I need to go.

Have a great day.

**Reply by Rooyca** [reply]
> do you suppose your code blocks could be made to render markdown?

You could use the html version form of the markdown directly on the `format`, no?

**Reply by julroger2013** [reply]
This works very well! Definitely suited for my purpose! Thanks for doing this.

Re using HTML versus rendering markdown ...

HTML looks better in the codeblock, but won't copy checkboxes when clicking the copy button on the codeblock. 

MD formatting copies properly and can then be pasted elsewhere in Obsidian without issue, but doesn't look pretty in its codeblock rendering.

Candidly, I would think this probably isn't worth your time to implement unless it's a very easy change. The important part (at least for me) is that I can use the "copy contents" function to grab the information I need from the codeblock now. 

Thanks again!

**Reply by Rooyca** [reply]
I don't know how hard it'll be but I'll try to implement it. :wink:

**Comment by julroger2013** [comment]
Awesome, thank you. Really great plugin - I appreciate all the time you’ve spent making it!Again, I suspect that Better Search Views would contain some code relevant to markdown rendering.You’re the best.On May 7, 2024, at 11:08, Ronald Cantillo ***@***.***> wrote:﻿
I don't know how hard it'll be but I'll try to implement it. 😉

**Reply by Rooyca** [reply]
Hey @julroger2013,
I believe is done! Please tell me if it works to you.

**Reply by julroger2013** [reply]
Yes, pretty much! It certainly looks neater.

The only slight oddity I'm noticing is that the markdown checkbox renders, but it is also preceded by a bullet point, as in this screenshot:

<img width="40" alt="image" src="https://github.com/Rooyca/obsidian-api-request/assets/94991208/6fce4f0c-1a05-4e5d-9b3b-833becd3db4a">

**Reply by Rooyca** [reply]
Yes, not only that it also doesn't works! When you click it nothing happens. I believe has something to do with [marked](https://marked.js.org/) (I'm using that for parsing the MD) but i wasn't able to find a solution. (I know the `input` is disabled but I don't know why).

I was trying to fix it but came to the conclusion that wasn't worth it because it will reset after you click on the codeblock. So didn't make much sense to me.

**Reply by julroger2013** [reply]
It worked perfect for me with the copy contents button! I was able to paste a beautifully formatted list of markdown checkboxes.

Are you referring to the checkboxes not getting checked off when clicked? I wouldn't necessarily expect them to.

**Reply by Rooyca** [reply]
Yes, when clicked in the codeblock. The copy button works nicely. 

Was thinking on adding  another button call "Render" or something like that which would render the codeblock as a `div` making it easy to work with the data. *(Not sure how to do that yet tho.)*

**Reply by Rooyca** [reply]
Hello @julroger2013,

I added a new flag (`req-id`) that stores the response in `localStorage`. Maybe that could help you with your workflow.

**Comment by julroger2013** [comment]
This is great. It's working really, really well.

If it's worth your time, I could see it being very useful having an API for your plugin. It's not really necessary for my purposes - the copy/paste into my Templater Script is plenty easy to do. 

But I can imagine some of the more JavaScript-savvy Obsidian users would love being able to directly plug the results of one of your API requests into something like Templater, where the result of the API request is just pulled as an input that can then be used within a larger script.

Thanks so much for all you have done! This plugin is really amazing.

**Reply by Rooyca** [reply]
That's exactly what I would like to achieve, the `req-id` flag is my first approach. 
I can't really think where to start, maybe I should look at examples of plugins that have that functionality. Any recommendations?
